### ISSUE TITLE
LSTM behaving differently with recurrent activation "sigmoid" and tf.keras.activations.sigmoid

### ISSUE THREAD

**Issue by tvatsal1996** [tensorflow/tensorflow#41028]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux 3.10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): pip install tensorflow
- TensorFlow version (use command below): 2.1.0
- Python version: 3.6.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.1/7.6.1
- GPU model and memory: Tesla V100-SXM2-32GB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
Using tf.keras.layers.LSTM(units, recurrent_activation=tf.keras.activations.sigmoid) - My training is not converging
Using tf.keras.layers.LSTM(units, recurrent_activation='sigmoid') - Same training is converging
Also the time taken in former is higher than the latter. This I have found that in recurrent_v2.py, it is checking if recurrent_activation == 'sigmoid' which fails in case of tf.keras.activations.sigmoid, so the former one is not using CuDNN while latter one is using. What is different in those two cases is that the latter one is converging using CuDNN while the former one is not converging at all not using CuDNN.

**Describe the expected behavior**
Training should converge for both of the above however training time variation is understood as former is not using CuDNN but the latter one is using.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.
No error is coming just training is acting differently

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by amahendrakar** [comment]
@tvatsal1996,
In order to expedite the trouble-shooting process, could you please provide the complete code to reproduce the issue reported here. Thanks!

**Comment by tvatsal1996** [comment]
Here is a sample code for my issue on the specified system details:
```
import tensorflow as tf
from CustomBatchGenerator import FileDataset

class TranNetwork(tf.keras.Model):
    """Transcription Network
    """
    def __init__(self, num_lstm_layers, lstm_cell_size, dropout=0.0):
        super(TranNetwork, self).__init__()

        self.num_lstm_layers = num_lstm_layers
        self.lstm_cell_size = lstm_cell_size
        self.dropout = dropout
        self.trans_layers = []
        self.pooling = tf.keras.layers.MaxPool1D(pool_size=3, padding="same")

        # HERE IS THE ISSUE, USING RECURRENT ACTIVATION AS tf.keras.activations.sigmoid instead of 'sigmoid'
        # causes model to not converge at all
        recurrent_activation = 'sigmoid'
        #recurrent_activation = tf.keras.activations.sigmoid
        for l in range(num_lstm_layers):
          self.trans_layers.append(tf.keras.layers.LSTM(
                                     self.lstm_cell_size,
                                     return_sequences=True,
                                     dropout=self.dropout,
                                     recurrent_activation=recurrent_activation))

    def call(self, x, x_len, training=True):
        seq_len = x_len
        output = tf.clip_by_value(x, -3.0, 3.0)
        for l in range(self.num_lstm_layers):
            mask = tf.sequence_mask(seq_len)
            output = self.trans_layers[l](output, mask=mask, training=training)
            if l == 0:
              seq_len = tf.cast(tf.math.ceil(tf.divide(seq_len, 3)), dtype=tf.int32)
              output = self.pooling(output, training=training)

        mask = tf.sequence_mask(seq_len)
        output = self.trans_layers[-1](output, mask=mask, training=training)

        return output, seq_len

class RNNTModel(object):
    """ RNNT Model class for training
    """
    def __init__(self, num_lstm_layers, lstm_cell_size, vocab_size, dropout=0.0):
        self.vocab_size = vocab_size
        self.trans = TranNetwork(num_lstm_layers, lstm_cell_size, dropout=dropout)
        self.ctc_layer = tf.keras.layers.Dense(1+self.vocab_size, name='ctc')

    def forward(self, x, y, x_len, y_len, training=True):
        @tf.function(input_signature=[
            tf.TensorSpec(shape=[None, None, 40], dtype=tf.float32),
            tf.TensorSpec(shape=[None, None], dtype=tf.int32),
            tf.TensorSpec(shape=[None, ], dtype=tf.int32),
            tf.TensorSpec(shape=[None, ], dtype=tf.int32),
            tf.TensorSpec(shape=[], dtype=tf.bool)])
        def _forward(x, y, x_len, y_len, training=True):
            trans_output, output_len = self.trans(x, x_len, training=training)
            ctc_out = self.ctc_layer(trans_output, training=training)
            output_d = {'ctc_out': ctc_out, 'output_len': output_len}
            return output_d
        return _forward(x, y, x_len, y_len, training=training)

    def loss(self, y, x_len, y_len, ctc_out):
        @tf.function(input_signature=[
            tf.TensorSpec(shape=[None, None], dtype=tf.int32),
            tf.TensorSpec(shape=[None, ], dtype=tf.int32),
            tf.TensorSpec(shape=[None, ], dtype=tf.int32),
            tf.TensorSpec(shape=[None, None, 1+self.vocab_size], dtype=tf.float32)])
        def _loss(y, x_len, y_len, ctc_out):
            ctc_loss = tf.nn.ctc_loss(y, ctc_out, y_len, x_len,
                                      logits_time_major=False,
                                      blank_index=self.vocab_size)
            # to ignore invalid ctc loss case
            mask = tf.dtypes.cast(
              tf.math.greater_equal(x_len, y_len), dtype=tf.float32)
            ctc_loss = tf.multiply(ctc_loss, mask)
            ctc_loss = tf.reduce_sum(ctc_loss)
            return ctc_loss
        return _loss(y, x_len, y_len, ctc_out)

    @property
    def trainable_variables(self):
        return self.trans.trainable_variables \
               + self.ctc_layer.trainable_variables

model = RNNTModel(2, 32, 29, dropout=0.0)
optimizer = tf.keras.optimizers.Adam(0.00025)

@tf.function(input_signature=[
    tf.TensorSpec(shape=[None, None, 40], dtype=tf.float32),
    tf.TensorSpec(shape=[None, None], dtype=tf.int32),
    tf.TensorSpec(shape=[None, ], dtype=tf.int32),
    tf.TensorSpec(shape=[None, ], dtype=tf.int32)])
def train_step(x, y, x_len, y_len):
    with tf.GradientTape() as tape:
        output_d = model.forward(x, y, x_len, y_len, training=True)
        ctc_loss = model.loss(y, output_d['output_len'],
                              y_len, output_d['ctc_out'])
    variables = model.trainable_variables
    gradients = tape.gradient(ctc_loss, variables)
    gradients, _ = tf.clip_by_global_norm(gradients, 1.0)
    optimizer.apply_gradients(zip(gradients, variables))
    loss_norm_factor = 1.0 / tf.cast(tf.reduce_sum(y_len), dtype=tf.float32)
    return ctc_loss * loss_norm_factor
```

**Comment by amahendrakar** [comment]
@tvatsal1996,
I did not face any error while running the given code with TF v2.2. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/e8109f044f1593a27bd1189197a112c4/41028.ipynb). Thanks!

**Comment by tvatsal1996** [comment]
> @tvatsal1996,
> I did not face any error while running the given code with TF v2.2. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/e8109f044f1593a27bd1189197a112c4/41028.ipynb).
> 
> Looks like the code doesn't reach the `recurrent_activation = tf.keras.activations.sigmoid` line. Could you please provide a reproducible example. Thanks!

Hi, it will not throw out any error. I was asking that these two models which are very much the same except recurrent activation is "sigmoid" in the first one and tf.keras.activations.sigmoid in the second one, but the still with tf.keras.activations.sigmoid, my model does not converge.

**Comment by tvatsal1996** [comment]
Hi, I am still stuck in this and have not got any workaround till now. Please help !!

**Comment by PKlumpp** [comment]
I was digging through the implementations of GRU and LSTM when I stumbled over your post. Maybe this is an explanation for the behavior you are observing:
In the implementation of recurrent_v2.py, you can choose both recurrent_activation and activation, but this seems to have no effect when using the standard_lstm and standard_gru. Note that for cuDNN you must not change the activations.
Here is the implementation:

https://github.com/tensorflow/tensorflow/blob/8c6f2d55762f3fc94f98fdd8b3c5d59ee1276dba/tensorflow/python/keras/layers/recurrent_v2.py#L1298

This was different in the older implementations, but recent versions seem to ignore the given activations. Or am I wrong?

**Comment by tvatsal1996** [comment]
Yes, recent version seems to ignore the given activations in [standard_lstm](https://github.com/tensorflow/tensorflow/blob/b36436b087bd8e8701ef51718179037cccdfc26e/tensorflow/python/keras/layers/recurrent_v2.py#L1238), but my issue is that my model is not converging when run without CuDNN kernel. When being run with CuDNN kernel, i.e, when using  [gpu_lstm](https://github.com/tensorflow/tensorflow/blob/b36436b087bd8e8701ef51718179037cccdfc26e/tensorflow/python/keras/layers/recurrent_v2.py#L1320), it was converging with recurrent activation as sigmoid but while not using CuDNN kernel, i.e when using [standard_lstm](https://github.com/tensorflow/tensorflow/blob/b36436b087bd8e8701ef51718179037cccdfc26e/tensorflow/python/keras/layers/recurrent_v2.py#L1238) function, it was not converging. Is there something I am missing ?

**Comment by qlzh727** [comment]
The check for cudnn kernel compatibility was updated a while ago to fix this case. See https://github.com/tensorflow/tensorflow/blob/05632ed9bad5bf9eee3edd57ade3d8250d580019/tensorflow/python/keras/layers/recurrent_v2.py#L1090. 

Could u try with the latest TF release?

**Comment by qlzh727** [comment]
> I was digging through the implementations of GRU and LSTM when I stumbled over your post. Maybe this is an explanation for the behavior you are observing:
> In the implementation of recurrent_v2.py, you can choose both recurrent_activation and activation, but this seems to have no effect when using the standard_lstm and standard_gru. Note that for cuDNN you must not change the activations.
> Here is the implementation:
> 
> https://github.com/tensorflow/tensorflow/blob/8c6f2d55762f3fc94f98fdd8b3c5d59ee1276dba/tensorflow/python/keras/layers/recurrent_v2.py#L1298
> 
> This was different in the older implementations, but recent versions seem to ignore the given activations. Or am I wrong?

Note that for standard_lstm/gpu_lstm, its the code path that _could_use_gpu_kernel = True (which means the value of activation and recurrent_activation should be tanh and sigmoid). The code path for _could_use_gpu_kernel = False is falling back to LSTM v1 which is in recurrent.py.

**Comment by qlzh727** [comment]
Since this is already fixed in the latest nightly, I am closing this bug now.
